src/solvers/LevenbergMarquardt.jl

```diff
@@ -0,0 +1,236 @@
+) where {mT<:AbstractManifold,TF,TJ}
+    return NonlinearLeastSquaresProblem{typeof(evaluation),mT,TF,TJ}(M, F, jacF)
+end
+
+function get_cost(P::NonlinearLeastSquaresProblem, p)
+    return norm(P.F(P.M, p))^2
+end
+
+@doc raw"""
+    GradientDescentOptions{P,T} <: AbstractGradientOptions
+
+Describes a Gradient based descent algorithm, with
+
+# Fields
+a default value is given in brackets if a parameter can be left out in initialization.
+
+* `x` – a point (of type `P`) on a manifold as starting point
+* `jacF` – the current Jacobian of ``F``
+* `stopping_criterion` – ([`StopAfterIteration`](@ref)`(100)`) a [`StoppingCriterion`](@ref)
+* `retraction_method` – (`default_retraction_method(M)`) the retraction to use, defaults to
+  the default set for your manifold.
+
+# Constructor
+
+    LevenbergMarquardtOptions(M, x; initial_vector=zero_vector(M, x), kwargs...)
+
+Generate gradient descent options, where `initial_vector` can be used to set the tangent vector to store the gradient to a certain type.
+All following fields are keyword arguments.
+
+# See also
+[`gradient_descent`](@ref), [`GradientProblem`](@ref)
+"""
+mutable struct LevenbergMarquardtOptions{
+    P,
+    TM<:AbstractManifold,
+    TStop<:StoppingCriterion,
+    TRTM<:AbstractRetractionMethod,
+    TJac,
+    Tparams<:Real,
+} <: AbstractGradientOptions
+    M::TM
+    x::P
+    stop::TStop
+    retraction_method::TRTM
+    jacF::TJac
+    η::Tparams
+    μ::Tparams
+    β::Tparams
+    flagnz::Bool
+    function LevenbergMarquardtOptions{P}(
+        M::AbstractManifold,
+        initialX::P,
+        initial_jacF::TJac,
+        s::StoppingCriterion=StopAfterIteration(100),
+        retraction_method::AbstractRetractionMethod=ExponentialRetraction(),
+        η::Real=0.2,
+        μmin::Real=0.1,
+        β::Real=5.0,
+        flagnz::Bool=false,
+    ) where {P,TJac}
+        if η <= 0 || η >= 1
+            throw(ArgumentError("Value of η must be strictly between 0 and 1, received $η"))
+        end
+        if μmin <= 0
+            throw(ArgumentError("Value of μmin must be strictly above 0, received $μmin"))
+        end
+        if β <= 1
+            throw(ArgumentError("Value of β must be strictly above 1, received $β"))
+        end
+        Tparams = promote_type(typeof(η), typeof(μmin), typeof(β))
+        return new{P,typeof(M),typeof(s),typeof(retraction_method),TJac,Tparams}(
+            M, initialX, s, retraction_method, initial_jacF, η, μmin, β, flagnz
+        )
+    end
+end
+
+function LevenbergMarquardtOptions(
+    M::AbstractManifold,
+    x::P,
+    initial_jacF::TJac;
+    stopping_criterion::StoppingCriterion=StopAfterIteration(100),
+    retraction_method::AbstractRetractionMethod=default_retraction_method(M),
+) where {P,TJac}
+    return LevenbergMarquardtOptions{P}(
+        M, x, initial_jacF, stopping_criterion, retraction_method
+    )
+end
+
+function get_gradient(o::LevenbergMarquardtOptions)
```

Comment:
The `Options` `o` should not store the manifold. Note that 

https://manoptjl.org/stable/plans/problem/#Manopt.get_gradient

has two meanings (maybe we should change that):

1. with options it returns the last evaluated gradient (stored within `O`)
2. with `Problem` `P` and a point `x` it evaluates the gradient stored / computed from the static part in P (that is the manifold and the cost function and the gradient function and so on).

Reply:
also Jack seems to be similar to the gradient a static/unmodified function/thing that should be stored in the Problem?

Reply:
Note that to implement `get_gradient` I need to either store gradient in options or compute it from what is stored there. And `step_solver!` doesn't explicitly compute a gradient, so I opted to store what is actually computed anyway (Jacobian of F) and compute gradient as needed.

Reply:
> also Jack seems to be similar to the gradient a static/unmodified function/thing that should be stored in the Problem?

Could you rephrase that? I'm not sure what you mean.

Reply:
> And step_solver! doesn't explicitly compute a gradient,
Or maybe it should, let me think about it.

Reply:
> Note that to implement `get_gradient` I need to either store gradient in options or compute it from what is stored there. And `step_solver!` doesn't explicitly compute a gradient, so I opted to store what is actually computed anyway (Jacobian of F) and compute gradient as needed.

The gradient should be able to be computed given the problem (containing the static data like the cost and all its derivatives/subgradiens/jacobians of inner vector stuff) and an iterate. You could use a function `get_jacobian` which you introduce and use in the solver, which could also be internally used by `get_gradient` then?

Reply:
I've removed manifold from options, in the end it turned out to be not needed.